src/qforte/utils/symmetry_analysis.py

```diff
@@ -0,0 +1,131 @@
+    N_exp_val = np.real(N_exp_val)
+    N_var = np.real(N_var)
+
+    # Compute <Sz> and <Sz^2> - <Sz>^2, where Sz is the projection of the total spin on the z axis
+    Sz = qf.total_spin_z(n_qubits)
+    Sz_sqrd = qf.QubitOperator()
+    Sz_sqrd.add_op(Sz)
+    Sz_sqrd.operator_product(Sz, True, True)
+    Sz_exp_val = qc.direct_op_exp_val(Sz)
+    Sz_var = qc.direct_op_exp_val(Sz_sqrd) - Sz_exp_val**2
+    if np.imag(Sz_exp_val) != 0 or np.imag(Sz_var) != 0:
+        raise ValueError('The expectation value and variance of Sz should be real!')
+    Sz_exp_val = np.real(Sz_exp_val)
+    Sz_var = np.real(Sz_var)
+
+    # Compute <S^2> and <S^4> - <S^2>^2, where S is the total spin
+    S_sqrd = qf.total_spin_squared(n_qubits)
+    S_sqrd_sqrd = qf.QubitOperator()
+    S_sqrd_sqrd.add_op(S_sqrd)
+    S_sqrd_sqrd.operator_product(S_sqrd, True, True)
+    S_sqrd_exp_val = qc.direct_op_exp_val(S_sqrd)
+    S_sqrd_var = qc.direct_op_exp_val(S_sqrd_sqrd) - S_sqrd_exp_val**2
+    if np.imag(S_sqrd_exp_val) !=0 or np.imag(S_sqrd_var) !=0:
+        raise ValueError('The expectation value and variance of S^2 should be real!')
+    S_sqrd_exp_val = np.real(S_sqrd_exp_val)
+    S_sqrd_var = np.real(S_sqrd_var)
+
+    # Compute the contributions of various symmetry spaces to the wavefunction
+    weight_particle_number = [0] * (n_qubits + 1)
+    weight_Sz = [0] * (n_qubits + 1)
+    Sz_values = []
+    Sz_values_dictionary = {}
+    for i in range(n_qubits + 1):
+        Sz_values.append(-n_qubits * 0.25 + 0.5 * i)
+        Sz_values_dictionary[Sz_values[i]] = i
+    weight_irreps = [0] * len(irreps)
+    for det in range(1<<n_qubits):
+        occ_alpha = []
+        occ_beta = []
+        for i in range(0, n_qubits, 2):
+            if (1<<i)&det != 0:
+                occ_alpha.append(i)
+            if (1<<(i+1))&det != 0:
+                occ_beta.append(i+1)
+        weight = qc.get_coeff_vec()[det]
+        weight = weight * np.conjugate(weight)
+        particle_number = len(occ_alpha + occ_beta)
+        Sz_value = (len(occ_alpha) - len(occ_beta)) * 0.5
+        irrep = qf.sq_op_find_symmetry(orb_irreps_to_int, occ_alpha + occ_beta, [])
+        weight_particle_number[particle_number] += weight
+        weight_Sz[Sz_values_dictionary[Sz_value]] += weight
+        weight_irreps[irrep] += weight
+    for i in range(n_qubits + 1):
+        weight_particle_number[i] = np.real(weight_particle_number[i])
+        weight_Sz[i] = np.real(weight_Sz[i])
+    for i in range(len(irreps)):
+        weight_irreps[i] = np.real(weight_irreps[i])
+
+    print('\n\n*******************')
+    print('*Symmetry Analysis*')
+    print('*******************')
+
+    print('\nParticle number symmetry analysis')
+    print('=================================')
+
+    for i in range(n_qubits + 1):
+        print('N =', str(i).rjust(2, ' '), 'weight:', f'{weight_particle_number[i]:.10f}')
```

Comment:
What are these "weights", and why do we print them on a per-qubit basis?

Reply:
The weights quantify the contribution of each symmetry sector to the final wavefunction. For example, let us assume that the wavefunction of a 6-electron (N = 6) system looks like this:

|Ψ> = c_1 |Φ_1> + c_2 |Φ_2> + c_3 |Φ_3> + c_4 |Φ_4> + c_5 |Φ_5> + c_6 |Φ_6> + d_1 |D_1> + d_2 |D_2>,

where  the last two determinants are contaminants containing N = 5 electrons. We thus have the following weights:

weight(N = 5) =  |d_1|^2 + |d_2|^2
weight(N = 6) =  |c_1|^2 + |c_2|^2 + |c_3|^2 + |c_4|^2 + |c_5|^2 + |c_6|^2

Normalization requires:

weight(N = 5) + weight(N = 6) = 1

They are not printed per qubit, but per total particle number, N in [0, n_qubits].

The code works in a similar manner for the Sz and spatial symmetries. Since we are using a basis of Slater determinants, it is not straightforward to do a similar analysis for S^2.